src/libs/Notification/PushNotification/subscribePushNotification/index.ts

```diff
@@ -50,27 +67,42 @@ function applyOnyxData({reportID, reportActionID, onyxData, lastUpdateID, previo
     }
 
     Log.info('[PushNotification] reliable onyx update received', false, {lastUpdateID, previousUpdateID, onyxDataCount: onyxData?.length ?? 0});
-    const updates: OnyxUpdatesFromServer = {
-        type: CONST.ONYX_UPDATE_TYPES.AIRSHIP,
-        lastUpdateID,
-        previousUpdateID,
-        updates: [
-            {
-                eventType: 'eventType',
-                data: onyxData,
-            },
-        ],
-    };
+
+    const updates = buildOnyxUpdatesFromServer({onyxData, lastUpdateID, previousUpdateID});
 
     /**
      * When this callback runs in the background on Android (via Headless JS), no other Onyx.connect callbacks will run. This means that
      * lastUpdateIDAppliedToClient will NOT be populated in other libs. To workaround this, we manually read the value here
      * and pass it as a param
      */
-    return getLastUpdateIDAppliedToClient().then((lastUpdateIDAppliedToClient) => applyOnyxUpdatesReliably(updates, true, lastUpdateIDAppliedToClient));
+    return getLastUpdateIDAppliedToClient().then(
+        (lastUpdateIDAppliedToClient) =>
+            new Promise((resolve) => {
+                applyOnyxUpdatesReliably(updates, true, lastUpdateIDAppliedToClient);
+                resolve();
+            }),
+    );
 }
 
 function navigateToReport({reportID, reportActionID}: ReportActionPushNotificationData): Promise<void> {
+    Log.info('[PushNotification] Adding push notification updates to deferred updates queue', false, {reportID, reportActionID});
+
+    // Onyx data from push notifications might not have been applied when they were received in the background
+    // due to OS limitations. So we'll also attempt to apply them here so they can display immediately. Reliable
+    // updates will prevent any old updates from being duplicated and any gaps in them will be handled
+    Airship.push.getActiveNotifications().then((notifications) => {
+        const onyxUpdates = notifications.map((notification) => getPushNotificationData(notification));
+        // const onyxData = onyxUpdates.reduce<OnyxServerUpdate[]>((onyxUpdatesAcc, onyxUpdate) => (onyxUpdate.onyxData ? [...onyxUpdatesAcc, ...onyxUpdate.onyxData] : onyxUpdatesAcc), []);
+
+        // const lastUpdateID = onyxUpdates.reduce((max, {lastUpdateID: lastUpdateIDIter}) => Math.max(max, lastUpdateIDIter ?? 0), 0);
+        // const previousUpdateID = onyxUpdates.find((update) => update.lastUpdateID === lastUpdateID)?.previousUpdateID ?? 0;
+
+        // const updates = wrapAirshipUpdatesDataInOnyxUpdatesFromServer({onyxData, lastUpdateID, previousUpdateID});
+        // return DeferredOnyxUpdates.enqueueAndProcess(updates);
+
+        return Promise.all(onyxUpdates.map((update) => applyOnyxData(update)));
```

Comment:
On second thought, I think we do need to `enqueueAndProcess` these updates as we're applying them in a random order here. 

The end goal we're looking for is that all of the push notification updates are sorted and then applied in-order. The DeferredOnyxUpdates queue should be able to do that with `enqueueAndProcess`. Does that make sense?

Reply:
The sorting and processing is actually also achieved by calling `applyOnyxData`. This will trigger `applyOnyxUpdatesReliably` which will then trigger `handleOnyxUpdateGap` and `DeferredOnyxUpdates.enqueue` functions consecutively.

Once the deferred updates are being processed by `validateAndApplyDeferredUpdates` they will be sorted and applied.

Reply:
gotcha! makes sense